aioaws/core.py

```diff
@@ -86,140 +189,84 @@ async def request(
-            'x-amz-date': _aws4_x_amz_date(now),
-        }
-
-        payload_sha256_hash = hashlib.sha256(data).hexdigest()
-        signed_headers, signature = self._aws4_signature(now, method, url, headers, payload_sha256_hash)
-        credential = self._aws4_credential(now)
-        authorization_header = (
-            f'{_AUTH_ALGORITHM} Credential={credential},SignedHeaders={signed_headers},Signature={signature}'
-        )
-        headers.update({'authorization': authorization_header, 'x-amz-content-sha256': payload_sha256_hash})
-        return headers
-
-    def _aws4_signature(
-        self, dt: datetime, method: Literal['GET', 'POST'], url: URL, headers: Dict[str, str], payload_hash: str
-    ) -> Tuple[str, str]:
-        header_keys = sorted(headers)
-        signed_headers = ';'.join(header_keys)
-        canonical_request_parts = (
-            method,
-            url_quote(url.path),
-            url.query.decode(),
-            ''.join(f'{k}:{headers[k]}\n' for k in header_keys),
-            signed_headers,
-            payload_hash,
-        )
-        canonical_request = '\n'.join(canonical_request_parts)
-        string_to_sign_parts = (
-            _AUTH_ALGORITHM,
-            _aws4_x_amz_date(dt),
-            self._aws4_scope(dt),
-            hashlib.sha256(canonical_request.encode()).hexdigest(),
-        )
-        string_to_sign = '\n'.join(string_to_sign_parts)
-        return signed_headers, self._aws4_sign_string(string_to_sign, dt)
-
-    def _aws4_sign_string(self, string_to_sign: str, dt: datetime) -> str:
-        key_parts = (
-            b'AWS4' + self.aws_secret_key.encode(),
-            _aws4_date_stamp(dt),
-            self.region,
-            self.service,
-            _AWS_AUTH_REQUEST,
-            string_to_sign,
-        )
-        signature_bytes: bytes = reduce(_aws4_reduce_signature, key_parts)  # type: ignore
-        return hexlify(signature_bytes).decode()
-
-    def _aws4_scope(self, dt: datetime) -> str:
-        return f'{_aws4_date_stamp(dt)}/{self.region}/{self.service}/{_AWS_AUTH_REQUEST}'
-
-    def _aws4_credential(self, dt: datetime) -> str:
-        return f'{self.aws_access_key}/{self._aws4_scope(dt)}'
-
-
-def _aws4_date_stamp(dt: datetime) -> str:
-    return dt.strftime('%Y%m%d')
-
-
-def _aws4_x_amz_date(dt: datetime) -> str:
-    return dt.strftime('%Y%m%dT%H%M%SZ')
-
-
-def _aws4_reduce_signature(key: bytes, msg: str) -> bytes:
-    return hmac.new(key, msg.encode(), hashlib.sha256).digest()
-
 
 class RequestError(RuntimeError):
-    def __init__(self, r: Response):
+    def __init__(self, r: httpx.Response):
         error_msg = f'unexpected response from {r.request.method} "{r.request.url}": {r.status_code}'
         super().__init__(error_msg)
         self.response = r
         self.status = r.status_code
 
     def __str__(self) -> str:
         return f'{self.args[0]}, response:\n{self.response.text}'
+
+
+class AWSV4AuthFlow(httpx.Auth):
```

Comment:
The `AWSClient` class was going to be difficult to adapt to SQS, e.g. because the host name for some requests depends on the queue URL while others depend on the service/region. So this is an attempt to decouple the auth portion of things from the hostname / other stuff going on in AWSClient. I can try to find a way to work around things instead if you'd prefer.